### ISSUE TITLE
fails to initialize additional stages when running in parallel under TestNG

### ISSUE THREAD

**Issue by adrian-herscu** [TNG/JGiven#300]
Given a scenario
With a `@ScenarioStage` annotated field
When running it with `org.testng.annotations.Test(singleThreaded = false)`
Then the field should be initialized

After changing to `singleThreaded = true` the field is initialized.

**Comment by adrian-herscu** [comment]
looks similar to #105

**Comment by janschaefer** [comment]
If I understand you correctly you have something like:

    class MyTest {
        @ScenarioStage
        MyStage stage;

        @Test
        public void mytest() {
            stage.given().something()
        }
    }

right?

**Comment by janschaefer** [comment]
I currently think that this problem cannot be fixed. It is an inherent problem of TestNG that the test methods share the same test instance (in contrast to JUnit). The only possibility that I see is that you add the stage within your test method. Something like:

    @Test
    public void mytest() {
          MyStage myStage = getScenario().addStage(MyStage.class);
          myStage.given().something();
    }

**Comment by adrian-herscu** [comment]
Yes.
Can you hold the scenario state in a ThreadLocal?
Just the first thought...
Adrian.

בתאריך 10 בפבר' 2017 11:22,‏ "Jan" <notifications@github.com> כתב:

If I understand you correctly you have something like:

class MyTest {
    @ScenarioStage
    MyStage stage;

    @Test
    public void mytest() {
        stage.given().something()
    }
}

right?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/TNG/JGiven/issues/300#issuecomment-278895898>, or mute
the thread

**Comment by janschaefer** [comment]
The problem is not that the scenario state itself is not thread-safe. It is more a TestNG internal problem. The only solution that might work would be to introduce some proxy instance that is injected into the TestNG test instances and that delegates to the actual stage instances of the corresponding thread.

**Comment by l-1squared** [comment]
superseded by #829
